### PR TITLE
Add $XDG_DATA_DIRS checks

### PR DESCRIPTION
**Pull request by DoomHammer**
- [X] Have you followed the guidelines in our [Contributing](https://github.com/Linuxbrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Linuxbrew/brew/pulls) for the same change?
- [X] Have you added an explanation of what your changes do and why you'd like us to include them?
- [X] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [X] Have you successfully run `brew tests` with your changes locally?

---

As found in https://github.com/Linuxbrew/legacy-linuxbrew/pull/919

### CODE REVIEW

**Comment by sjackman** [comment]
Do you know whether this issue also affects Mac? Should this diagnostic be run on both OS or just Linux?

**Comment by DoomHammer** [comment]
Cannot tell lacking access to a Mac. My guess would be it should not be needed there as Mac should probably have some other means of discovery than following XDG. On the other hand, Cross Desktop may stretch further than open-source desktops it was originally developed for.

File: Library/Homebrew/diagnostic.rb [hunk 1]
```diff
@@ -535,6 +535,24 @@ def check_user_path_3
         EOS
       end
 
+      def check_xdg_data_dirs
+        share = "#{HOMEBREW_PREFIX}/share"
+        homebrew_in_xdg_data_dirs = true
+        if ENV.key?("XDG_DATA_DIRS")
+          xdg_data_dirs = ENV["XDG_DATA_DIRS"].split(File::PATH_SEPARATOR)
+          xdg_data_dirs.length > 0 && homebrew_in_xdg_data_dirs = xdg_data_dirs.include?(share)
```
Comment (sjackman) [inline Library/Homebrew/diagnostic.rb]: Avoid this style of using `&&` and assignment.

Comment (sjackman) [inline Library/Homebrew/diagnostic.rb]: Try (untested code)

``` ruby
homebrew_in_xdg_data_dirs = !ENV["XDG_DATA_DIRS"] ||
    ENV["XDG_DATA_DIRS"].split(File::PATH_SEPARATOR).include?(share)
```

Comment (DoomHammer) [inline Library/Homebrew/diagnostic.rb]: Unfortunately, this won't do. Replaced `&&` with `if`.

Comment (sjackman) [inline Library/Homebrew/diagnostic.rb]: I'm okay with that.

For my curiosity, which case didn't my code handle? Are you concerned about the case where `$XDG_DATA_DIRS` is set to the empty string?

Comment (DoomHammer) [inline Library/Homebrew/diagnostic.rb]: Yup, that one.

Comment (sjackman) [inline Library/Homebrew/diagnostic.rb]: That should probably give a warning though, yeah? I believe that case works with my code.

Comment (DoomHammer) [inline Library/Homebrew/diagnostic.rb]: One of the tests tests (sic!) this case. When the variable is set but it is empty we get the warning which should not be there as `vapigen` will work with it.

Your approach is more concise, then. It does no harm, it simply is more strict than it needs to be.

Comment (sjackman) [inline Library/Homebrew/diagnostic.rb]: I'm partial to the more concise code, and giving the warning when `XDG_DATA_DIRS` is set to the empty string.

**Comment by DoomHammer** [comment]
How about now, @sjackman ? Concise (mostly) _and_ correct.

File: Library/Homebrew/diagnostic.rb [hunk 2]
```diff
@@ -516,6 +516,22 @@ def check_user_path_3
...
+        homebrew_in_xdg_data_dirs = !ENV["XDG_DATA_DIRS"] ||
+          ENV["XDG_DATA_DIRS"].split(File::PATH_SEPARATOR).include?(share) ||
+          ENV["XDG_DATA_DIRS"] == ""
+        unless homebrew_in_xdg_data_dirs
+          <<-EOS.undent
+          Homebrew's share was not found in your XDG_DATA_DIRS but you have
+          this variable set to include other locations.
+          Some programs like `vapigen` may not work correctly.
+          Consider setting the XDG_DATA_DIRS for example like so
+              echo 'export XDG_DATA_DIRS="#{share}:$XDG_DATA_DIRS"' >> #{shell_profile}
```
Comment (bw-yuhogg) [inline Library/Homebrew/diagnostic.rb:530]: Do we also want to put something like this into the installation script?

**Comment by DoomHammer** [comment]
I have to say the errors from CI puzzle me.

File: Library/Homebrew/diagnostic.rb (hunk 2 above)
Comment (DoomHammer) [reply Library/Homebrew/diagnostic.rb:530]: Might be a good idea. I hope the logic is quite clear? Unset XDG_DATA_DIRS or the one containing Homebrew `share` are considered good. Set but lacking Homebrew `share` is bad.

**Comment by sjackman** [comment]
Travis:

```
Error: Testing requires the latest version of testball_copy
```

https://travis-ci.org/Linuxbrew/brew/builds/154865066#L222

Circle:

```
1) Failure:
IntegrationCommandTests#test_test_formula [/home/ubuntu/brew/Library/Homebrew/test/test_integration_cmds.rb:844]:
Expected /This\ command\ requires\ a\ formula\ argument/
 to match "Error: cannot load such file -- test/unit/assertions\nPlease report this bug:\n    https://github.com/Linuxbrew/brew/blob/master/share/doc/homebrew/Troubleshooting.md#troubleshooting\n/home/ubuntu/brew/Library/Homebrew/formula_assertions.rb:7:in `require'\n/home/ubuntu/brew/Library/Homebrew/formula_assertions.rb:7:in `rescue in <module:Assertions>'\n/home/ubuntu/brew/Library/Homebrew/formula_assertions.rb:4:in `<module:Assertions>'\n/home/ubuntu/brew/Library/Homebrew/formula_assertions.rb:2:in `<module:Homebrew>'\n/home/ubuntu/brew/Library/Homebrew/formula_assertions.rb:1:in `<top (required)>'\n/home/ubuntu/brew/Library/Homebrew/cmd/test.rb:19:in `require'\n/home/ubuntu/brew/Library/Homebrew/cmd/test.rb:19:in `<top (required)>'\n/home/ubuntu/brew/Library/Homebrew/brew.rb:22:in `require'\n/home/ubuntu/brew/Library/Homebrew/brew.rb:22:in `require?'\n/home/ubuntu/brew/Library/Homebrew/brew.rb:61:in `<main>'"
```

This CI failure is unrelated to this PR and fixed by d2bbafa

File: Library/Homebrew/diagnostic.rb [hunk 3]
```diff
@@ -516,6 +516,22 @@ def check_user_path_3
         EOS
       end
 
+      def check_xdg_data_dirs
+        share = "#{HOMEBREW_PREFIX}/share"
+        homebrew_in_xdg_data_dirs = !ENV["XDG_DATA_DIRS"] ||
+          ENV["XDG_DATA_DIRS"].split(File::PATH_SEPARATOR).include?(share) ||
+          ENV["XDG_DATA_DIRS"] == ""
```
Comment (sjackman) [inline Library/Homebrew/diagnostic.rb]: A very small thing: I'd swap lines 522 and 523.

``` ruby
 homebrew_in_xdg_data_dirs = ENV["XDG_DATA_DIRS"].nil? || ENV["XDG_DATA_DIRS"] == "" ||
    ENV["XDG_DATA_DIRS"].split(File::PATH_SEPARATOR).include?(share)
```

Comment (DoomHammer) [reply Library/Homebrew/diagnostic.rb]: Done that. Should I merge then?

**Comment by sjackman** [comment]
Yep, go ahead and merge, Piotr. Thanks!

**Pull request closed without merging**